packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -65,10 +67,16 @@ export interface IPendingDetachedContainerState extends SnapshotWithBlobs {
 	pendingRuntimeState?: unknown;
 }
 
+interface SnapshotInfo extends SnapshotWithBlobs {
+	snapshotSequenceNumber: number;
+}
+
 export class SerializedStateManager {
-	private readonly processedOps: ISequencedDocumentMessage[] = [];
+	private processedOps: ISequencedDocumentMessage[] = [];
 	private snapshot: SnapshotWithBlobs | undefined;
 	private readonly mc: MonitoringContext;
+	private latestSnapshot: SnapshotInfo | undefined;
+	private refreshSnapshotLock: boolean = false;
```

Comment:
I think it would be better to track the promise itself, rather than a Boolean. An idea to make this even safer would be to create an immutable LazyPromise, that does the refresh `private readonly LazyPromise<SnapshotInfo>(...)`. The lazy promise won't start running until the promise is used, await, then, catch, etc

Basically all of refreshSnapshot and refreshPendingAttributes would be inlined in the lazy promises function callback. This has the added advantage that the methods are no longer callable, since they are only supported to be called once, and in a the current specific flow. 

Then in fetchSnapshot, we would start the promise by using it, like you setup the catch currently.

Reply:
refreshPendingAttributes could be called more than once, for the most likely scenario that we're behind the snapshot and need to wait. I still think all what you said stands for refreshSnapshot.